ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java

```diff
@@ -3377,6 +3382,37 @@ private final boolean checkIsCurrent ()
         return isCurrent;
     }
 
+    // called by the reaper system via TwoPhaseCoordinator.recordStackTraces
+    protected synchronized void recordStackTraces() {
+        Map<String,String> currentCapture = createStackTraces();
+        StackTraceCapture stackTraceCapture = new StackTraceCapture(System.currentTimeMillis(), currentCapture);
+        threadStackTraceHistoryList.add(stackTraceCapture);
+    }
+
+    // called by the reaper system via TwoPhaseCoordinator.outputCapturedStackTraces
+    protected synchronized void outputCapturedStackTraces() {
+        DateFormat dateFormat = new SimpleDateFormat("HH:mm:ss.SSS");
+        for(StackTraceCapture stackTraceCapture : threadStackTraceHistoryList) {
+            for(Map.Entry<String,String> entry : createStackTraces().entrySet()) {
+                tsLogger.i18NLogger.info_historic_stack_trace(objectUid, entry.getKey(), dateFormat.format(new Date(stackTraceCapture.getTime())), entry.getValue());
```

Comment:
Is it fine to use it as logger level INFO?

Reply:
Support (and me) don't want it at DEBUG, because many users are reluctant to enable debug logging in production due to (perceived or actual) slowness. One option I did toy with is having a separate logger just for this feature, which means it can be independently set to the desired level and also that it can write only the stacktraces to a different file from other logging, which would potentially allow stack captures to be logged immediately instead of held in memory until the end of the tx (which is done so they appear grouped in the log for easer reading), so then you would get info even if the tx did not timeout but just committed more slowly than intended.  However, for the way it's currently designed, INFO is intended.

Reply:
I see, thanks for the clarification.